src/atl/formula/parser.rs

```diff
@@ -0,0 +1,653 @@
+        until,
+    };
+    use crate::atl::formula::Phi;
+
+    fn convert(id: String) -> Result<usize, Error> {
+        Ok(0)
+    }
+
+    struct Error {}
+
+    impl Debug for Error {
+        fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+            unimplemented!()
+        }
+    }
+
+    #[test]
+    fn paren_1() {
+        assert_eq!(paren(&convert).parse(b"(true)"), Ok(Phi::True))
+    }
+
+    fn convert_int(test: String) -> Result<usize, std::num::ParseIntError> {
+        test.parse()
+    }
+
+    /// Test for a single player
+    #[test]
+    fn enforce_players_1() {
+        assert_eq!(
+            enforce_players(&convert_int).parse(b"<<1>>"),
+            Ok(vec![1usize])
+        )
+    }
+
+    /// Test for two players
+    #[test]
+    fn enforce_players_2() {
+        assert_eq!(
+            enforce_players(&convert_int).parse(b"<<4,9>>"),
+            Ok(vec![4usize, 9usize])
+        )
+    }
+
+    /// Test for three players
+    #[test]
+    fn enforce_players_3() {
+        assert_eq!(
+            enforce_players(&convert_int).parse(b"<<203,23,4>>"),
+            Ok(vec![203usize, 23usize, 4usize])
+        )
+    }
+
+    /// The list of players is allowed to have whitespace after the seperator
+    #[test]
+    fn enforce_players_4() {
+        assert_eq!(
+            enforce_players(&convert_int).parse(b"<<203,  23>>"),
+            Ok(vec![203usize, 23usize])
+        )
+    }
+
+    /// Test for no players, should be valid
+    #[test]
+    fn enforce_players_5() {
+        assert_eq!(enforce_players(&convert_int).parse(b"<<>>"), Ok(vec![]))
+    }
+
+    /// Test for a single player
+    #[test]
+    fn despite_players_1() {
+        assert_eq!(
+            despite_players(&convert_int).parse(b"[[1]]"),
+            Ok(vec![1usize])
+        )
+    }
+
+    /// Test for two players
+    #[test]
+    fn despite_players_2() {
+        assert_eq!(
+            despite_players(&convert_int).parse(b"[[4,9]]"),
+            Ok(vec![4usize, 9usize])
+        )
+    }
+
+    /// Test for three players
+    #[test]
+    fn despite_players_3() {
+        assert_eq!(
+            despite_players(&convert_int).parse(b"[[203,23,4]]"),
+            Ok(vec![203usize, 23usize, 4usize])
+        )
+    }
+
+    /// The list of players is allowed to have whitespace after the seperator
+    #[test]
+    fn despite_players_4() {
+        assert_eq!(
+            despite_players(&convert_int).parse(b"[[203,  23]]"),
+            Ok(vec![203usize, 23usize])
+        )
+    }
+
+    /// Test for no players, should be valid
+    #[test]
+    fn despite_players_5() {
+        assert_eq!(despite_players(&convert_int).parse(b"[[]]"), Ok(vec![]))
+    }
+
+    #[test]
+    fn next_1() {
+        assert_eq!(next(&convert).parse(b"X true"), Ok(Phi::True))
+    }
+
+    #[test]
+    fn next_2() {
+        assert_eq!(next(&convert).parse(b"Xtrue"), Ok(Phi::True))
```

Comment:
I am not sure this should be allowed. Same goes for the other operators. But it's a small thing.